internal/controllers/core/filewatch/controller.go

```diff
@@ -15,31 +15,93 @@ package filewatch
 import (
 	"context"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 
-	"github.com/tilt-dev/tilt/internal/store"
 	filewatches "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+	"github.com/tilt-dev/tilt/pkg/logger"
 )
 
-// FileWatchController reconciles a FileWatch object
+// fsFileWatchFinalizer cleans up the actual filesystem (fs) monitor for a FileWatch spec.
+const fsFileWatchFinalizer = "fswatch.filewatch.finalizers.core.tilt.dev"
+
+// Controller reconciles a filewatches.FileWatch object.
 type Controller struct {
 	ctrlclient.Client
-	Store store.RStore
+
+	WatchManager *ApiServerWatchManager
 }
 
-func NewController(store store.RStore) *Controller {
+func NewController(wm *ApiServerWatchManager) *Controller {
 	return &Controller{
-		Store: store,
+		WatchManager: wm,
 	}
 }
 
 // +kubebuilder:rbac:groups=,resources=filewatches,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=,resources=filewatches/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=,resources=filewatches/finalizers,verbs=update
 
-func (r *Controller) Reconcile(_ context.Context, _ ctrl.Request) (ctrl.Result, error) {
-	// this is currently a no-op stub
+func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := logger.Get(ctx).WithFields(logger.Fields{"apiserver_entity": req.NamespacedName.String()})
+	ctx = logger.WithLogger(ctx, log)
+
+	var fileWatchApiObj filewatches.FileWatch
+	if err := r.Get(ctx, req.NamespacedName, &fileWatchApiObj); err != nil {
+		return ctrl.Result{}, ctrlclient.IgnoreNotFound(err)
+	}
+
+	if fileWatchApiObj.ObjectMeta.DeletionTimestamp.IsZero() {
+		// ensure finalizer is attached to non-deleted objects so that the actual filesystem-level
+		// monitor is removed upon deletion
+		if !controllerutil.ContainsFinalizer(&fileWatchApiObj, fsFileWatchFinalizer) {
+			controllerutil.AddFinalizer(&fileWatchApiObj, fsFileWatchFinalizer)
+			if err := r.Update(ctx, &fileWatchApiObj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if controllerutil.ContainsFinalizer(&fileWatchApiObj, fsFileWatchFinalizer) {
+			if err := r.WatchManager.StopWatch(fileWatchApiObj.Name); err != nil {
```

Comment:
i wonder if finalizers are overkill for what we're doing here...they're designed for external resources that persist even if the controller server reboots, but that's obviously not the case with this particular controller.

Reply:
That's fair, without a finalizer, it could be handled when the object isn't found earlier in `Reconcile()`, especially since we intentionally _don't_ want to block actual deletion even if something goes awry with the file watcher.

Reply:
Got rid of the finalizer along with dramatically simplifying implementation